src/sinks/util/unix.rs

```diff
@@ -136,131 +142,58 @@ impl UnixSink {
+            debug!(
+                message = "Connecting",
+                path = %self.path.to_str().unwrap()
+            );
+            match UnixStream::connect(self.path.clone()).await {
+                Ok(stream) => {
+                    emit!(UnixSocketConnectionEstablished { path: &self.path });
+                    return AckerBytesSink::new(
+                        stream,
+                        self.acker.clone(),
+                        Box::new(|byte_size| emit!(UnixSocketEventSent { byte_size })),
                     );
-                    let fut = self.connector.connect();
-                    UnixSinkState::Creating(Box::new(fut.compat()))
                 }
-            }
-        }
-    }
-}
-
-impl Sink for UnixSink {
-    type SinkItem = Bytes;
-    type SinkError = ();
-
-    fn start_send(&mut self, line: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        let byte_size = line.len();
-        match self.poll_connection() {
-            Ok(Async::Ready(connection)) => match connection.start_send(line) {
                 Err(error) => {
-                    emit!(UnixSocketSendFailed {
+                    emit!(UnixSocketConnectionFailure {
                         error,
-                        path: &self.connector.path,
+                        path: &self.path
                     });
-                    self.state = UnixSinkState::Disconnected;
-                    Ok(AsyncSink::Ready)
-                }
-                Ok(res) => {
-                    emit!(UnixSocketEventSent { byte_size });
-                    Ok(res)
+                    delay_for(backoff.next().unwrap()).await;
                 }
-            },
-            Ok(Async::NotReady) => Ok(AsyncSink::NotReady(line)),
-            Err(_) => unreachable!(),
+            }
         }
     }
+}
 
-    fn poll_complete(&mut self) -> Poll01<(), Self::SinkError> {
-        // Stream::forward will immediately poll_complete the sink it's forwarding to,
-        // but we don't want to connect before the first event actually comes through.
-        if let UnixSinkState::Disconnected = self.state {
-            return Ok(Async::Ready(()));
-        }
-
-        let connection = try_ready!(self.poll_connection());
-
-        match connection.poll_complete() {
-            Err(error) => {
-                emit!(UnixSocketFlushFailed {
+#[async_trait]
+impl StreamSink for UnixSink {
+    async fn run(&mut self, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let encoding = self.encoding.clone();
+        let mut input = input
+            .map(|event| match encode_event(event, &encoding) {
+                Some(bytes) => bytes,
+                None => Bytes::new(),
+            })
+            .map(Ok)
+            .peekable();
+
+        while Pin::new(&mut input).peek().await.is_some() {
+            let mut sink = self.connect().await;
+            if let Err(error) = sink.send_all(&mut input).await {
```

Comment:
If I'm understanding this correctly, on every event we will initiate a new connection with a new sink stack and flush that one event all the way though? Doing things one at a time like this will likely be extremely slow and does not get us any benefit from batching writes or acks.

Reply:
No, it's not correct. It was in original PR (https://github.com/timberio/vector/pull/3893). Here we initiate connection once we have an item in the stream (`.peek()`), but not consume it. Once we have an item we connect and forward all items from the stream. If we have error on sending, we log it and wait for new item again, then start the connection, and so on.

Reply:
Ah, that makes sense. Sorry, should have read closer! :smile: